hw/net/rocker/rocker_p4_rmt.c

```diff
@@ -0,0 +1,216 @@
+/*
+ * QEMU rocker switch emulation - PCI device - P4 support
+ *
+ * Copyright (c) 2015 Parag Bhide <parag.bhide@barefootnetworks.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ */
+/*
+ * Common functionality that can be leveraged by all P4 worlds
+ */
+#include <stdarg.h>
+#include "net/eth.h"
+#include "qemu/iov.h"
+#include "qemu/timer.h"
+#include "qmp-commands.h"
+
+#include "rocker.h"
+#include "rocker_hw.h"
+#include "rocker_fp.h"
+#include "rocker_tlv.h"
+#include "rocker_world.h"
+#include "rocker_desc.h"
+#include "rocker_p4_rmt.h"
+#include "rocker_p4_l2l3_features.h"
+#include "rocker_p4_l2l3_enums.h"
+#include "rocker_p4_l2l3_tables.h"
+#include "rocker_p4_l2l3_tables_if.h"
+#include "rocker_p4_l2l3.h"
+#include "rocker_p4_common.h"
+
+unsigned char *rocker_p4_iov_to_buffer(const struct iovec *iov, int iovcnt);
+static int rocker_p4_rmt_table_cmd (struct p4_rmt_world *p4_rmt, 
+                                    struct desc_info *info,
+                                    char *buf, unsigned short cmd,
+                                    RockerTlv **tlvs);
+
+unsigned char *
+rocker_p4_iov_to_buffer(const struct iovec *iov, int iovcnt)
+{
+    // convert iov to a single contiguous buffer for p4 model
+    int len = 0;
+    unsigned char *buf = NULL;
+
+    len = iov_size(iov, iovcnt);
+    DPRINTF("convert %d iovs to a buffer of len %d\n", iovcnt, len);
+    buf = malloc(len);
```

Comment:
This was done on purpose.. (But I need to fix it) as the P4 behavior model
frees the packet buffer using "free" not g_free.
But the model code does not allocate the buffer.. I'll abstract it by
creating pkt_buf_alloc() function that will be provided by the model
library.

On Sun, Apr 19, 2015 at 10:26 PM, Scott Feldman <notifications@github.com>
wrote:

> In hw/net/rocker/rocker_p4_rmt.c
> <https://github.com/pbhide/qemu-rocker/pull/1#discussion_r28664658>:
>
> > +unsigned char *rocker_p4_iov_to_buffer(const struct iovec *iov, int iovcnt);
> > +static int rocker_p4_rmt_table_cmd (struct p4_rmt_world *p4_rmt,
> > +                                    struct desc_info *info,
> > +                                    char *buf, unsigned short cmd,
> > +                                    RockerTlv **tlvs);
> > +
> > +unsigned char *
> > +rocker_p4_iov_to_buffer(const struct iovec *iov, int iovcnt)
> > +{
> > +    // convert iov to a single contiguous buffer for p4 model
> > +    int len = 0;
> > +    unsigned char *buf = NULL;
> > +
> > +    len = iov_size(iov, iovcnt);
> > +    DPRINTF("convert %d iovs to a buffer of len %d\n", iovcnt, len);
> > +    buf = malloc(len);
>
> Use g_malloc(), g_free()
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pbhide/qemu-rocker/pull/1/files#r28664658>.
>